File: apps/agent-manager-gui/src/api/marketplace.ts
import { invoke } from "@tauri-apps/api/core";

import type { Platform } from "../types";

export type MarketplaceSort = "installs" | "stars" | "newest";

export interface MarketplaceSkill {
  slug: string;
  name: string;
  source: string;
  summary: string;
  github_stars: number;
  installs: number;
  github_language?: string | null;
  categories: string[];
}

export interface MarketplaceListResult {
  skills: MarketplaceSkill[];
  total: number;
  offset: number;
  limit: number;
}

export interface SkillAddBatchItem {
  source: string;
  skillName?: string;
}

export interface SkillAddBatchOutcome {
  ok: number;
  failed: number;
  errors: string[];
}

export function fetchMarketplaceSkills(params: {
  sort: MarketplaceSort;
  q?: string;
  sourceFilter?: string;
  offset?: number;
  limit?: number;
}): Promise<MarketplaceListResult> {
  return invoke<MarketplaceListResult>("cmd_marketplace_list_skills", {
    sort: params.sort,
    q: params.q ?? null,
    sourceFilter: params.sourceFilter ?? null,
    offset: params.offset ?? 0,
    limit: params.limit ?? 50,
  });
}

export function addSkillsBatch(
  project: string,
  skills: SkillAddBatchItem[],
  targetPlatforms: Platform[],
): Promise<SkillAddBatchOutcome> {
  return invoke<SkillAddBatchOutcome>("cmd_skill_add_batch", {
    project,
    skills: skills.map((s) => ({
      source: s.source,
      skill_name: s.skillName ?? null,
    })),
    targetPlatforms,
  });
}

export function skillRowKey(skill: MarketplaceSkill): string {
  return `${skill.source}/${skill.name}`;
}

export function formatMarketplaceCount(n: number): string {
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`;
  if (n >= 1_000) return `${(n / 1_000).toFixed(1)}k`;
  return String(n);
}

File: apps/agent-manager-gui/src/marketplace.rs
//! [Claude Code Marketplace](https://www.claudemarketplace.net/skills) 列表 API 代理。
//!
//! Tauri WebView 无法直接跨域 fetch，故由 Rust 侧请求 `GET /api/skills`。

use std::time::Duration;

use serde::{Deserialize, Serialize};

const API_BASE: &str = "https://www.claudemarketplace.net/api/skills";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MarketplaceSkill {
    pub slug: String,
    pub name: String,
    pub source: String,
    #[serde(default)]
    pub summary: String,
    #[serde(default)]
    pub github_stars: u64,
    #[serde(default)]
    pub installs: u64,
    pub github_language: Option<String>,
    #[serde(default)]
    pub categories: Vec<String>,
}

#[derive(Debug, Serialize)]
pub struct MarketplaceListResult {
    pub skills: Vec<MarketplaceSkill>,
    pub total: u64,
    pub offset: u32,
    pub limit: u32,
}

#[derive(Deserialize)]
struct ApiResponse {
    skills: Vec<MarketplaceSkill>,
    total: u64,
}

pub fn list_skills(
    sort: &str,
    q: Option<&str>,
    source_filter: Option<&str>,
    offset: u32,
    limit: u32,
) -> Result<MarketplaceListResult, String> {
    let sort = match sort {
        "stars" | "newest" | "installs" => sort,
        _ => "installs",
    };
    let limit = limit.clamp(1, 100);

    let client = reqwest::blocking::Client::builder()
        .timeout(Duration::from_secs(45))
        .build()
        .map_err(|e| format!("HTTP 客户端初始化失败: {e}"))?;

    let mut req = client.get(API_BASE).query(&[
        ("sort", sort),
        ("offset", &offset.to_string()),
        ("limit", &limit.to_string()),
    ]);

    if let Some(query) = q.filter(|s| !s.trim().is_empty()) {
        req = req.query(&[("q", query.trim())]);
    }

    let resp = req
        .send()
        .map_err(|e| format!("请求 Claude Marketplace 失败: {e}"))?;

    if !resp.status().is_success() {
        return Err(format!(
            "Claude Marketplace 返回 HTTP {}",
            resp.status().as_u16()
        ));
    }

    let mut data: ApiResponse = resp
        .json()
        .map_err(|e| format!("解析 Marketplace JSON 失败: {e}"))?;

    if let Some(src) = source_filter.filter(|s| !s.trim().is_empty()) {
        let src = src.trim();
        data.skills.retain(|s| s.source == src);
    }

    Ok(MarketplaceListResult {
        skills: data.skills,
        total: data.total,
        offset,
        limit,
    })
}
